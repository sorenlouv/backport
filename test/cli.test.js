const inquirer = require('inquirer');
const axios = require('axios');
const nock = require('nock');
const httpAdapter = require('axios/lib/adapters/http');
const { init } = require('../src/cli');
const github = require('../src/github');
const commitsMock = require('./mocks/commits.json');
const { mockBackportDirPath } = require('./testHelpers');
const utils = require('../src/utils');

function setup() {
  const owner = 'elastic';
  const repoName = 'backport-cli-test';
  const fullRepoName = `${owner}/${repoName}`;
  axios.defaults.host = 'http://localhost';
  axios.defaults.adapter = httpAdapter;

  mockBackportDirPath();
  utils.exec = jest.fn().mockReturnValue(Promise.resolve());
  utils.writeFile = jest.fn().mockReturnValue(Promise.resolve());
  github.getCommits = jest.fn(github.getCommits);
  github.createPullRequest = jest.fn(github.createPullRequest);

  inquirer.prompt = jest
    .fn()
    .mockReturnValueOnce(Promise.resolve({ promptResult: fullRepoName }))
    .mockReturnValueOnce(
      Promise.resolve({
        promptResult: {
          message: 'myCommitMessage',
          sha: 'mySha'
        }
      })
    )
    .mockReturnValueOnce(
      Promise.resolve({
        promptResult: '6.2'
      })
    );

  nock('https://api.github.com')
    .get(`/repos/${owner}/${repoName}/commits`)
    .query({ author: 'sqren', per_page: '5', access_token: 'myAccessToken' })
    .reply(200, commitsMock);

  nock('https://api.github.com')
    .get(`/search/issues`)
    .query({
      q: 'repo:elastic/backport-cli-test mySha',
      access_token: 'myAccessToken'
    })
    .reply(200, {
      items: [
        {
          number: 'myPullRequest'
        }
      ]
    });

  nock('https://api.github.com')
    .post(`/repos/${owner}/${repoName}/pulls`)
    .query({ access_token: 'myAccessToken' })
    .reply(200, {
      html_url: 'myHtmlUrl'
    });

  return init(
    {
      username: 'sqren',
      accessToken: 'myAccessToken',
      repositories: [
        {
          name: fullRepoName,
          versions: ['6.x', '6.0', '5.6', '5.5', '5.4']
        }
      ]
    },
    { cwd: '/my/path', own: true, multiple: false }
  );
}

describe('select commit that originated from pull request', () => {
  beforeEach(setup);

  it('getCommit should be called with correct args', () => {
    expect(github.getCommits).toHaveBeenCalledWith(
      'elastic',
      'backport-cli-test',
      'sqren'
    );
  });

  it('createPullRequest should be called with correct args', () => {
    expect(github.createPullRequest).toHaveBeenCalledWith(
      'elastic',
      'backport-cli-test',
      {
        base: '6.2',
        body: 'Backports pull request #myPullRequest to 6.2',
        head: 'sqren:backport/6.2/pr-myPullRequest',
        title: '[6.2] myCommitMessage'
      }
    );
  });

  it('prompt calls should match snapshot', () => {
    expect(inquirer.prompt.mock.calls).toMatchSnapshot();
  });

  it('exec should be called with correct args', () => {
    expect(utils.exec.mock.calls).toMatchSnapshot();
  });

  it('writeFile should be called with correct args', () => {
    expect(utils.writeFile.mock.calls).toMatchSnapshot();
  });
});
