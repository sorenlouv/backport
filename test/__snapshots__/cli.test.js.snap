// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select commit that originated from commit exec should be called with correct args 1`] = `
Array [
  Array [
    "git clone git@github.com:elastic/backport-cli-test",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic",
    },
  ],
  Array [
    "git remote add sqren git@github.com:sqren/backport-cli-test",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git reset --hard && git checkout master && git pull origin master",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git fetch origin 6.2 && git branch backport/6.2/commit-mySha origin/6.2 --force && git checkout backport/6.2/commit-mySha ",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git push sqren backport/6.2/commit-mySha --force",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
]
`;

exports[`select commit that originated from commit prompt calls should match snapshot 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          "elastic/backport-cli-test",
        ],
        "message": "Select repository",
        "name": "fullRepoName",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "date": "2017-09-05T08:48:42Z",
              "message": "Update readme: Make ui-framework url a link",
              "sha": "7939a98aadb94ead30b54bea7974cd02dc816b02",
            },
          },
          Object {
            "name": "Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "date": "2017-08-28T08:27:51Z",
              "message": "Update react 15.6.1 (#13672)",
              "sha": "52552a93efa5b4038ea57d9cb3d158ba6a1ff71d",
            },
          },
          Object {
            "name": "Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "date": "2017-08-07T15:35:59Z",
              "message": "Mock out static files when running in Jest (#13315)",
              "sha": "8cfbb82cfb50aa962410a9682d96335a252d751a",
            },
          },
        ],
        "message": "Select commit to backport",
        "name": "commit",
        "pageSize": 3,
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "6.x",
          "6.0",
          "5.6",
          "5.5",
          "5.4",
        ],
        "message": "Select version to backport to",
        "name": "versions",
        "type": "checkbox",
      },
    ],
  ],
]
`;

exports[`select commit that originated from commit writeFile should be called with correct args 1`] = `
Array [
  Array [
    "homefolder/.backport/config.json",
    "{
  // Github personal access token. Must be created here: https://github.com/settings/tokens/new
  // Must have \\"Repo: Full control of private repositories\\"
  \\"accessToken\\": \\"\\",

  // Github username, eg. kimchy
  \\"username\\": \\"\\",

  // Only allow picking own commits to backport
  \\"own\\": true,

  // Allow picking multiple versions to backporting to
  \\"multiple\\": true,

  // Repositories and the versions that will be available in backport cli
  \\"repositories\\": [
    {
      \\"name\\": \\"elastic/x-pack-kibana\\",
      \\"versions\\": [\\"6.x\\", \\"6.0\\", \\"5.6\\", \\"5.5\\", \\"5.4\\"]
    },
    {
      \\"name\\": \\"elastic/kibana\\",
      \\"versions\\": [\\"6.x\\", \\"6.0\\", \\"5.6\\", \\"5.5\\", \\"5.4\\"]
    }
  ]
}
",
    Object {
      "flag": "wx",
    },
  ],
]
`;

exports[`select commit that originated from pull request exec should be called with correct args 1`] = `
Array [
  Array [
    "git clone git@github.com:elastic/backport-cli-test",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic",
    },
  ],
  Array [
    "git remote add sqren git@github.com:sqren/backport-cli-test",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git reset --hard && git checkout master && git pull origin master",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git fetch origin 6.2 && git branch backport/6.2/pr-myPullRequest origin/6.2 --force && git checkout backport/6.2/pr-myPullRequest ",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git cherry-pick mySha",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
  Array [
    "git push sqren backport/6.2/pr-myPullRequest --force",
    Object {
      "cwd": "homefolder/.backport/repositories/elastic/backport-cli-test",
    },
  ],
]
`;

exports[`select commit that originated from pull request prompt calls should match snapshot 1`] = `
Array [
  Array [
    Array [
      Object {
        "choices": Array [
          "elastic/backport-cli-test",
        ],
        "message": "Select repository",
        "name": "fullRepoName",
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          Object {
            "name": "Update readme: Make ui-framework url a link",
            "short": "Update readme: Make ui-framework url a link",
            "value": Object {
              "date": "2017-09-05T08:48:42Z",
              "message": "Update readme: Make ui-framework url a link",
              "sha": "7939a98aadb94ead30b54bea7974cd02dc816b02",
            },
          },
          Object {
            "name": "Update react 15.6.1 (#13672)",
            "short": "Update react 15.6.1 (#13672)",
            "value": Object {
              "date": "2017-08-28T08:27:51Z",
              "message": "Update react 15.6.1 (#13672)",
              "sha": "52552a93efa5b4038ea57d9cb3d158ba6a1ff71d",
            },
          },
          Object {
            "name": "Mock out static files when running in Jest (#13315)",
            "short": "Mock out static files when running in Jest (#13315)",
            "value": Object {
              "date": "2017-08-07T15:35:59Z",
              "message": "Mock out static files when running in Jest (#13315)",
              "sha": "8cfbb82cfb50aa962410a9682d96335a252d751a",
            },
          },
        ],
        "message": "Select commit to backport",
        "name": "commit",
        "pageSize": 3,
        "type": "list",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "choices": Array [
          "6.x",
          "6.0",
          "5.6",
          "5.5",
          "5.4",
        ],
        "message": "Select version to backport to",
        "name": "version",
        "type": "list",
      },
    ],
  ],
]
`;

exports[`select commit that originated from pull request writeFile should be called with correct args 1`] = `
Array [
  Array [
    "homefolder/.backport/config.json",
    "{
  // Github personal access token. Must be created here: https://github.com/settings/tokens/new
  // Must have \\"Repo: Full control of private repositories\\"
  \\"accessToken\\": \\"\\",

  // Github username, eg. kimchy
  \\"username\\": \\"\\",

  // Only allow picking own commits to backport
  \\"own\\": true,

  // Allow picking multiple versions to backporting to
  \\"multiple\\": true,

  // Repositories and the versions that will be available in backport cli
  \\"repositories\\": [
    {
      \\"name\\": \\"elastic/x-pack-kibana\\",
      \\"versions\\": [\\"6.x\\", \\"6.0\\", \\"5.6\\", \\"5.5\\", \\"5.4\\"]
    },
    {
      \\"name\\": \\"elastic/kibana\\",
      \\"versions\\": [\\"6.x\\", \\"6.0\\", \\"5.6\\", \\"5.5\\", \\"5.4\\"]
    }
  ]
}
",
    Object {
      "flag": "wx",
    },
  ],
]
`;
